### PR TITLE
Support creating and dropping olap table with StarOSTablet

### PR DESCRIPTION
**Pull request by wyb**
## What type of PR is this：
- [ ] bug
- [x] feature
- [ ] enhancement
- [ ] others

## Problem Summary(Required) ：

1. Support creating olap table with StarOSTablet when the user specifies
   the S3 storage medium.
2. Support dropping olap table with StarOSTablet.
3. Add a new class StarOSAgent to encapsulate StarClient interface.
4. Mock StarClient and StarOS worker.

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:disappointed: **fail** : 90 / 123 (73.17%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/CatalogRecycleBin.java|0|10|00.00%|
|:large_blue_circle:|com/starrocks/master/MasterImpl.java|3|4|75.00%|
|:large_blue_circle:|com/starrocks/catalog/Catalog.java|50|65|76.92%|
|:large_blue_circle:|com/starrocks/catalog/StarOSAgent.java|30|37|81.08%|
|:large_blue_circle:|com/starrocks/common/util/PropertyAnalyzer.java|2|2|100.00%|
|:large_blue_circle:|com/starrocks/catalog/StarOSTablet.java|1|1|100.00%|
|:large_blue_circle:|com/starrocks/catalog/MaterializedIndex.java|2|2|100.00%|
|:large_blue_circle:|com/starrocks/common/Config.java|1|1|100.00%|
|:large_blue_circle:|com/starrocks/catalog/TabletMeta.java|1|1|100.00%|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:disappointed: **fail** : 91 / 115 (79.13%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/CatalogRecycleBin.java|2|10|20.00%|
|:large_blue_circle:|com/starrocks/master/MasterImpl.java|3|4|75.00%|
|:large_blue_circle:|com/starrocks/catalog/Catalog.java|50|65|76.92%|
|:large_blue_circle:|com/starrocks/common/util/PropertyAnalyzer.java|2|2|100.00%|
|:large_blue_circle:|com/starrocks/catalog/StarOSTablet.java|1|1|100.00%|
|:large_blue_circle:|com/starrocks/catalog/MaterializedIndex.java|2|2|100.00%|
|:large_blue_circle:|com/starrocks/common/Config.java|1|1|100.00%|
|:large_blue_circle:|com/starrocks/catalog/StarOSAgent.java|29|29|100.00%|
|:large_blue_circle:|com/starrocks/catalog/TabletMeta.java|1|1|100.00%|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:heart_eyes: **pass** : 94 / 115 (81.74%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/CatalogRecycleBin.java|5|10|50.00%|
|:large_blue_circle:|com/starrocks/master/MasterImpl.java|3|4|75.00%|
|:large_blue_circle:|com/starrocks/catalog/Catalog.java|50|65|76.92%|
|:large_blue_circle:|com/starrocks/common/util/PropertyAnalyzer.java|2|2|100.00%|
|:large_blue_circle:|com/starrocks/catalog/StarOSTablet.java|1|1|100.00%|
|:large_blue_circle:|com/starrocks/catalog/MaterializedIndex.java|2|2|100.00%|
|:large_blue_circle:|com/starrocks/common/Config.java|1|1|100.00%|
|:large_blue_circle:|com/starrocks/catalog/StarOSAgent.java|29|29|100.00%|
|:large_blue_circle:|com/starrocks/catalog/TabletMeta.java|1|1|100.00%|

File: fe/fe-core/src/main/java/com/starrocks/catalog/Catalog.java [hunk 1]
```diff
@@ -4788,7 +4829,21 @@ private void createTablets(String clusterName, MaterializedIndex index, ReplicaS
...
 
         DistributionInfoType distributionInfoType = distributionInfo.getType();
-        if (distributionInfoType == DistributionInfoType.HASH) {
+        if (distributionInfoType != DistributionInfoType.HASH) {
+            throw new DdlException("Unknown distribution type: " + distributionInfoType);
+        }
+
+        if (tabletMeta.isUseStarOS()) {
+            // TODO: support colocate table
+            int bucketNum = distributionInfo.getBucketNum();
```
Comment (wuleistarrocks) [inline fe/fe-core/src/main/java/com/starrocks/catalog/Catalog.java:4838]: I have no idea about the difference between bucketNum and tablet id, can you explain it clearly?

File: fe/fe-core/src/main/java/com/starrocks/master/MasterImpl.java [hunk 2]
```diff
@@ -309,10 +309,10 @@ private void finishCreateReplica(AgentTask task, TFinishTaskRequest request) {
                         task.getBackendId() + ": " + request.getTask_status().getError_msgs().toString());
```
Comment (wuleistarrocks) [inline fe/fe-core/src/main/java/com/starrocks/master/MasterImpl.java:309]: if tablet create failed, we should consider delete the shard for staros tablet. I suggest add a TODO comment for this.

File: fe/fe-core/src/main/java/com/starrocks/catalog/MaterializedIndex.java [hunk 3]
```diff
@@ -219,6 +219,10 @@ public long getDataSize() {
     }
 
     public long getReplicaCount() {
+        if (useStarOS) {
```
[the remark this answers is not in the document]
Comment (wyb) [reply fe/fe-core/src/main/java/com/starrocks/catalog/MaterializedIndex.java:222]: For StarOSTablet, there are only replicas for computation, so i think we can use replica when processing query and not rely on replica in other places.

File: fe/fe-core/src/main/java/com/starrocks/master/MasterImpl.java (hunk 2 above)
Comment (wyb) [reply fe/fe-core/src/main/java/com/starrocks/master/MasterImpl.java:309]: Ok, i add a TODO first.

**Comment by wyb** [comment]
run starrocks_fe_unittest

**Comment by wyb** [comment]
run starrocks_fe_unittest

**Comment by wangruin** [comment]
run starrocks_fe_unittest

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:heart_eyes: **pass** : 94 / 115 (81.74%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/CatalogRecycleBin.java|5|10|50.00%|
|:large_blue_circle:|com/starrocks/master/MasterImpl.java|3|4|75.00%|
|:large_blue_circle:|com/starrocks/catalog/Catalog.java|50|65|76.92%|
|:large_blue_circle:|com/starrocks/common/util/PropertyAnalyzer.java|2|2|100.00%|
|:large_blue_circle:|com/starrocks/catalog/StarOSTablet.java|1|1|100.00%|
|:large_blue_circle:|com/starrocks/catalog/MaterializedIndex.java|2|2|100.00%|
|:large_blue_circle:|com/starrocks/common/Config.java|1|1|100.00%|
|:large_blue_circle:|com/starrocks/catalog/StarOSAgent.java|29|29|100.00%|
|:large_blue_circle:|com/starrocks/catalog/TabletMeta.java|1|1|100.00%|

File: fe/fe-core/src/main/java/com/starrocks/catalog/Catalog.java [hunk 4]
```diff
@@ -7509,12 +7561,18 @@ public void onEraseOlapTable(OlapTable olapTable, boolean isReplay) {
...
+                        if (partition.isUseStarOS()) {
+                            DropReplicaTask dropTask = new DropReplicaTask(
+                                    ((StarOSTablet) tablet).getPrimaryBackendId(), tabletId, schemaHash, true);
                             batchTask.addTask(dropTask);
-                        } // end for replicas
+                        } else {
+                            List<Replica> replicas = ((LocalTablet) tablet).getReplicas();
+                            for (Replica replica : replicas) {
+                                long backendId = replica.getBackendId();
+                                DropReplicaTask dropTask = new DropReplicaTask(backendId, tabletId, schemaHash, true);
```
Comment (gengjun-git) [inline fe/fe-core/src/main/java/com/starrocks/catalog/Catalog.java:7572]: How to distinguish the staros tablet drop and the local tablet drop On be?

**Review by gengjun-git — DISMISSED** [review]

File: fe/fe-core/src/main/java/com/starrocks/catalog/Catalog.java [hunk 5]
```diff
@@ -3796,11 +3809,12 @@ private void buildPartitionsConcurrently(long dbId, OlapTable table, List<Partit
...
                                                             MaterializedIndex index) {
         List<CreateReplicaTask> tasks = new ArrayList<>((int) index.getReplicaCount());
+        boolean useStarOS = partition.isUseStarOS();
         MaterializedIndexMeta indexMeta = table.getIndexMetaByIndexId(index.getId());
         for (Tablet tablet : index.getTablets()) {
-            for (Replica replica : ((LocalTablet) tablet).getReplicas()) {
+            if (useStarOS) {
                 CreateReplicaTask task = new CreateReplicaTask(
-                        replica.getBackendId(),
+                        ((StarOSTablet) tablet).getPrimaryBackendId(),
```
[the remark this answers is not in the document]
Comment (wyb) [reply fe/fe-core/src/main/java/com/starrocks/catalog/Catalog.java:3817]: Distinguish by storage medium and shard id.
I will add shard id later.

File: fe/fe-core/src/main/java/com/starrocks/catalog/Catalog.java (hunk 4 above)
Comment (wyb) [reply fe/fe-core/src/main/java/com/starrocks/catalog/Catalog.java:7572]: Distinguish by storage medium and shard id.
I will add shard id later.

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:heart_eyes: **pass** : 94 / 115 (81.74%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/CatalogRecycleBin.java|5|10|50.00%|
|:large_blue_circle:|com/starrocks/master/MasterImpl.java|3|4|75.00%|
|:large_blue_circle:|com/starrocks/catalog/Catalog.java|50|65|76.92%|
|:large_blue_circle:|com/starrocks/common/util/PropertyAnalyzer.java|2|2|100.00%|
|:large_blue_circle:|com/starrocks/catalog/StarOSTablet.java|1|1|100.00%|
|:large_blue_circle:|com/starrocks/catalog/MaterializedIndex.java|2|2|100.00%|
|:large_blue_circle:|com/starrocks/common/Config.java|1|1|100.00%|
|:large_blue_circle:|com/starrocks/catalog/StarOSAgent.java|29|29|100.00%|
|:large_blue_circle:|com/starrocks/catalog/TabletMeta.java|1|1|100.00%|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:heart_eyes: **pass** : 94 / 115 (81.74%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/CatalogRecycleBin.java|5|10|50.00%|
|:large_blue_circle:|com/starrocks/master/MasterImpl.java|3|4|75.00%|
|:large_blue_circle:|com/starrocks/catalog/Catalog.java|50|65|76.92%|
|:large_blue_circle:|com/starrocks/common/util/PropertyAnalyzer.java|2|2|100.00%|
|:large_blue_circle:|com/starrocks/catalog/StarOSTablet.java|1|1|100.00%|
|:large_blue_circle:|com/starrocks/catalog/MaterializedIndex.java|2|2|100.00%|
|:large_blue_circle:|com/starrocks/common/Config.java|1|1|100.00%|
|:large_blue_circle:|com/starrocks/catalog/StarOSAgent.java|29|29|100.00%|
|:large_blue_circle:|com/starrocks/catalog/TabletMeta.java|1|1|100.00%|

**Review by sduzh — APPROVED** [review]

**Review by gengjun-git — APPROVED** [review]

**Pull request merged**